src/ImageProcessor.Web/Caching/DiskCache.cs

```diff
@@ -207,46 +200,56 @@ public DiskCache(string requestPath, string fullPath, string querystring)
-                            // We'll cleanup any orphaned expired files though.
-                            if (!this.IsExpired(fileInfo.CreationTimeUtc) && count <= MaxFilesCount - 1)
-                            {
-                                break;
-                            }
-
-                            // Remove from the cache and delete each CachedImage.
-                            CacheIndexer.Remove(fileInfo.Name);
-                            fileInfo.Delete();
-                            count -= 1;
-                        }
-
-                        // ReSharper disable once EmptyGeneralCatchClause
-                        catch
-                        {
-                            // Log it but skip to the next file.
-                            ImageProcessorBootstrapper.Instance.Logger.Log<DiskCache>($"Unable to clean cached file: {fileInfo.FullName}");
-                        }
-                    }
-                }
+                return;
             }
+
+            await this.DebounceTrimmerAsync(async () =>
+             {
+                 string directory = Path.GetDirectoryName(this.CachedPath);
+
+                 if (directory != null)
+                 {
+                     // Jump up to the parent branch to clean through the cache.
+                     // ReSharper disable once PossibleNullReferenceException
+                     string parent = this.FolderDepth > 0 ? Path.GetFileName(this.CachedPath).Substring(0, 1) : string.Empty;
+                     DirectoryInfo rootDirectoryInfo = new DirectoryInfo(Path.Combine(validatedAbsoluteCachePath, parent));
+
+                     // UNC folders can throw exceptions if the file doesn't exist.
+                     foreach (DirectoryInfo enumerateDirectory in await rootDirectoryInfo.SafeEnumerateDirectoriesAsync())
+                     {
+                         IEnumerable<FileInfo> files = enumerateDirectory.EnumerateFiles().OrderBy(f => f.CreationTimeUtc);
+                         int count = files.Count();
+
+                         foreach (FileInfo fileInfo in files)
+                         {
+                             try
+                             {
+                                 // If the group count is equal to the max count minus 1 then we know we
+                                 // have reduced the number of items below the maximum allowed.
+                                 // We'll cleanup any orphaned expired files though.
+                                 if (!this.IsExpired(fileInfo.CreationTimeUtc) && count <= MaxFilesCount - 1)
+                                 {
+                                     break;
+                                 }
+
+                                 // Remove from the cache and delete each CachedImage.
+                                 CacheIndexer.Remove(fileInfo.Name);
+                                 fileInfo.Delete();
```

Comment:
In this logic are you ever checking if the folder is empty and therefore deleting the folder too? This would be especially useful for FCN since ASP.NET's monitoring is per folder.